File: hw/jacome/U2/HW12Cluster0/MicrophoneDB/src/main/java/ec/edu/espe/microphonedb/MicrophoneDB.java
package ec.edu.espe.microphonedb;

/**
 *
 * @author Micaela Jacome DESKTOP-46VMNHU ESPE
 */



import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;
import java.util.Scanner;

public class MicrophoneDB {
    static String connectionString = "mongodb+srv://Micaela:Micaela@cluster0.7ny5ynf.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
    
    public static void insertMicrophone() {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Microphone ID: ");
        String id = scanner.nextLine();

        System.out.print("Brand: ");
        String brand = scanner.nextLine();
        
        System.out.print("Model: ");
        String model = scanner.nextLine();
        
        System.out.print("Type: ");
        String type = scanner.nextLine();
        
        System.out.print("Color: ");
        String color = scanner.nextLine();

        System.out.print("Connection: ");
        String connection = scanner.nextLine();

        System.out.print("Frequency: ");
        String frequency = scanner.nextLine();

        System.out.print("Sensitivity: ");
        String sensitivity = scanner.nextLine();

        System.out.print("Impedance: ");
        String impedance = scanner.nextLine();

        System.out.print("Price: ");
        String price = scanner.nextLine();

        try (MongoClient mongoClient = MongoClients.create(connectionString)) {
            MongoDatabase database = mongoClient.getDatabase("Microphone");
            MongoCollection<Document> collection = database.getCollection("microphone");

            Document document = new Document("id", id)
                    .append("brand", brand)
                    .append("model", model)
                    .append("type", type)
                    .append("color", color)
                    .append("connection", connection)
                    .append("frequency", frequency)
                    .append("sensitivity", sensitivity)
                    .append("impedance", impedance)
                    .append("price", price);

            collection.insertOne(document);
            System.out.println("Microphone added successfully!");
        } catch (Exception e) {
            System.err.println("Error connecting to MongoDB: " + e.getMessage());
        }
    }

    public static void showMicrophones() {
        try (MongoClient mongoClient = MongoClients.create(connectionString)) {
            MongoDatabase database = mongoClient.getDatabase("Microphone");
            MongoCollection<Document> collection = database.getCollection("microphone");

            System.out.println("\nMicrophones in database:");

            for (Document doc : collection.find()) {
                System.out.println("ID: " + doc.getString("id"));
                System.out.println("Brand: " + doc.getString("brand"));
                System.out.println("Model: " + doc.getString("model"));
                System.out.println("Type: " + doc.getString("type"));
                System.out.println("Color: " + doc.getString("color"));
                System.out.println("Connection: " + doc.getString("connection"));
                System.out.println("Frequency: " + doc.getString("frequency"));
                System.out.println("Sensitivity: " + doc.getString("sensitivity"));
                System.out.println("Impedance: " + doc.getString("impedance"));
                System.out.println("Price: " + doc.getString("price"));
                System.out.println("-----------------------------");
            }
        } catch (Exception e) {
            System.err.println("Error connecting to MongoDB: " + e.getMessage());
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
      
        int option;
        do {
            System.out.println("\nMenu:");
            System.out.println("1. Add Microphone");
            System.out.println("2. Show Microphones");
            System.out.println("0. Exit");
            System.out.print("Select an option: ");
            option = scanner.nextInt();
            scanner.nextLine();

            switch (option) {
                case 1:
                    insertMicrophone();
                    break;
                case 2:
                    showMicrophones();
                    break;
                case 0:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid option, please try again.");
            }
        } while (option != 0);

        scanner.close();
    }
}

File: hw/jacome/U2/HW12Cluster0/JS/MicrophoneDB.js
const { MongoClient } = require('mongodb');
const readline = require('readline');

// Connection configuration
const connectionString = 'mongodb+srv://Micaela:Micaela@cluster0.7ny5ynf.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0ñ';
const dbName = 'Microphone';
const collectionName = 'microphone';

// Console interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Main function
async function main() {
  const client = new MongoClient(connectionString);
  
  try {
    await client.connect();
    console.log('Connected to MongoDB Atlas');
    
    const database = client.db(dbName);
    const collection = database.collection(collectionName);
    
    let option;
    do {
      console.log('\n=== MICROPHONE MANAGEMENT SYSTEM ===');
      console.log('1. Add microphone');
      console.log('2. List all microphones');
      console.log('3. Find microphone by ID');
      console.log('4. Update microphone');
      console.log('5. Delete microphone');
      console.log('0. Exit');
      
      option = await askQuestion('Select an option: ');
      
      switch (option) {
        case '1':
          await addMicrophone(collection);
          break;
        case '2':
          await listMicrophones(collection);
          break;
        case '3':
          await findMicrophoneById(collection);
          break;
        case '4':
          await updateMicrophone(collection);
          break;
        case '5':
          await deleteMicrophone(collection);
          break;
        case '0':
          console.log('Exiting system...');
          break;
        default:
          console.log('Invalid option');
      }
    } while (option !== '0');
    
  } catch (error) {
    console.error('Error:', error);
  } finally {
    await client.close();
    rl.close();
  }
}

// Helper functions
function askQuestion(question) {
  return new Promise(resolve => {
    rl.question(question, resolve);
  });
}

async function addMicrophone(collection) {
  console.log('\n--- ADD NEW MICROPHONE ---');
  
  const microphone = {
    id: await askQuestion('Microphone ID (e.g. 001): '),
    brand: await askQuestion('Brand (e.g. AudioTech): '),
    model: await askQuestion('Model (e.g. AT-1000): '),
    type: await askQuestion('Type (e.g. Condenser): '),
    color: await askQuestion('Color (e.g. Black): '),
    connection: await askQuestion('Connection (e.g. XLR): '),
    frequency: await askQuestion('Frequency (e.g. 20Hz-20kHz): '),
    sensitivity: await askQuestion('Sensitivity (e.g. -36dB): '),
    impedance: await askQuestion('Impedance (e.g. 250 ohms): '),
    price: await askQuestion('Price (e.g. 99.99): ')
  };
  
  const result = await collection.insertOne(microphone);
  console.log(`Microphone added with ID: ${result.insertedId}`);
}

async function listMicrophones(collection) {
  console.log('\n--- MICROPHONE LIST ---');
  const microphones = await collection.find().toArray();
  
  if (microphones.length === 0) {
    console.log('No microphones registered');
    return;
  }
  
  microphones.forEach(mic => {
    console.log(`
ID: ${mic.id}
Brand: ${mic.brand}
Model: ${mic.model}
Type: ${mic.type}
Color: ${mic.color}
Connection: ${mic.connection}
Frequency: ${mic.frequency}
Sensitivity: ${mic.sensitivity}
Impedance: ${mic.impedance}
Price: $${mic.price}
-----------------------------`);
  });
}

async function findMicrophoneById(collection) {
  const id = await askQuestion('Enter microphone ID to search: ');
  const mic = await collection.findOne({ id });
  
  if (mic) {
    console.log(`
=== MICROPHONE DETAILS ===
ID: ${mic.id}
Brand: ${mic.brand}
Model: ${mic.model}
Type: ${mic.type}
Color: ${mic.color}
Connection: ${mic.connection}
Frequency: ${mic.frequency}
Sensitivity: ${mic.sensitivity}
Impedance: ${mic.impedance}
Price: $${mic.price}`);
  } else {
    console.log('No microphone found with that ID');
  }
}

async function updateMicrophone(collection) {
  const id = await askQuestion('Enter microphone ID to update: ');
  
  const updates = {};
  const fields = ['brand', 'model', 'type', 'color', 'connection', 'frequency', 'sensitivity', 'impedance', 'price'];
  
  for (const field of fields) {
    const value = await askQuestion(`New value for ${field} (leave empty to keep current): `);
    if (value) updates[field] = value;
  }
  
  if (Object.keys(updates).length === 0) {
    console.log('No changes made');
    return;
  }
  
  const result = await collection.updateOne({ id }, { $set: updates });
  
  if (result.modifiedCount > 0) {
    console.log('Microphone updated successfully');
  } else {
    console.log('No microphone found with that ID');
  }
}

async function deleteMicrophone(collection) {
  const id = await askQuestion('Enter microphone ID to delete: ');
  const result = await collection.deleteOne({ id });
  
  if (result.deletedCount > 0) {
    console.log('Microphone deleted successfully');
  } else {
    console.log('No microphone found with that ID');
  }
}

// Start application
main().catch(console.error);